README
Information about all of the functions can be found in the Documentation folder

If you want to add code, please be sure to document it in the same style as the other functions are documented

Please try to stick to the Model-View-Controller architecture

ChangeLog.txt
R9(master, David's Branch, Codename: #ComeAtMeBro)
		New gameState, MENUOPEN.  This is the gamestate if you click on the object that is already clicked.  It opens a menu just like in advance wars.
		New class, UnitMenu.  This is a class containing information about menus that are drawn on the screen.
			X & Y grid position to draw it in (although this could (and probably should) have the option of drawing it in sub-grid space)
			Booleans to control which options are available
		New draw function: drawUnitMenu().  This draws the current UnitMenu
		New var in drawer.js: drawUnitMenu.  This is a boolean controlling if a UnitMenu is being drawn
		New var in drawer.js: unitMenu.  This is the UnitMenu that's being drawn (if there is one)
		New function in main.js: unitMenuOpen().  Returns true if the current state is UNITMENUOPEN
		I added a few other things also.  Not sure what, but nothing huge huge
		New class, menuoption, which contains the x, y, width, and height of a menu item, and its function.
		New function in gamelogic.js: menuClicked().  Tests to see if a menu option was clicked.
		Modified handleMouseClick() to support menu clicking
		New function in gamelogic.js: findMenuOption().  Returns the menu option object at (x, y)
		New function in unitmenu.js: menuOption.action(): Does whatever action the menu option is supposed to do
		Added new attribute, speed, to unit.  This controls the number of spaces it can move per turn
	TODO:
		Pathfinding
		Maximum move distance display
		Combat
	

R8(Quick Fixes)
	Changed some of the keyboard shit (you couldn't enter addunit state until you had selected something)
	Checks !occupied before addingunit

R7(And what?):
All sorts of shit up in here:
	!!!!!!Basically every single x,y you will see in any file is grid based. Even the drawer doesn't make a variable for a unit's x,y pixel position!!!!!!! #alwaysrememberthis
	addUnit to 1,1 rather than 32,32 //because of above
	deleted gx,gy from unit. x and y becoming grid based. Pixel value shouldn't be used.
	renamed moverelative to movepixelrelative //gets deleted in R7
	gridmoverelative to moverelative
	same with absolute versions of those.
	renamed validgridmove to validmove
	made playerMove tidier
	moveUnit in playerMove changed to moveUnitRelative 
	//need to make that function. #DONE
	moved validMoveRelative from gamelogic to unit. It wouldn't of worked if used it was used.
	drawSprites to * (x,y) of unit by gridpixel
	changed addunit,moveunit,setgridsquare,emptygridsquare a bit.
	moveunitrelative added. now calls the function IN the unit to move the units x,y position.
	Same with moveunit. 
	Added playerMoveAbsolute.
	deleting movepixelrelative - it's stupid. should only move by grid.
	deletingmovepixelabsolute - it's even more stupid. Somehow.
	added maxX and maxY variables which are set depending on size of canvas. makes some code easier to read and enables rectangular grid.
	can change moveable unit now. variable controlledUnit = 0, increased by 'a' and decreased by 's'. Stays within bounds of unitList.length. 
	can change moveable unit by clicking on it. I'm fucking amazing.
	added event listener for mouseup. The stolen part gets docx and docy which are the document coords of mouse click. the stolen function returns the position of an element (canvas). docx-elemx and docy-elemy give the coords relative to canvas.
	handlemouseclick converts pixel square clicked to grid square clicked.
	gridclicked finds the unit you clicked on from grid square clicked, if it exists and is clickable (added this bool to unit) calls clickedunit. #BREAKINGOUTOFANAMEDFORLOOPLIKEABOSS
	clickedunit sets controlledunit to the index of clicked unit. Also sorts out setnotcurrent and setcurrent // if (isyourturn) or something could go here #IVEDEALTWITHTHIS
	set
	Current and setNotCurrent added to unit. changes a boolean added to unit appropriately and changes sprite.
	Initial setCurrent is in Run function after adding the first unit. This is not legit.
	Added gameState array with YOURTURN and ADDUNIT
	changeState added 
	setNoCurrent sets there to be no current option
	showOccupied changed to showNotOccupied 
	In ADDUNIT state blue is shown where units can be added.
	click on a gridsquare to add a unit there
	yourTurn() and addingUnit() return true if in that state for convenience
	keyboard controls check state (no current controlled unit in addunit state, so would error hard)
	
	- Thought about making gridPixelx and gridPixely. To enable rectangular gridsquares. would involve changing sprite size depending on these values, which is long. Would make the grid system more awesome though if stolen for another game.
		#don'tstealthatshit
	- Thought about renaming moverelative/absolute to setxy or something, because it sets the x and y values, doesn't do anything else. Could be called by mistake thinking it does what playerMove does.
		#what'syouropinion.
	- Noticed EVERYTHING has relative and absolute versions. Although useful, it makes everything a bit less obvious/messy. We could stick to using one type only and delete the other, or continue as we are, or a third more badass option you need to think of.
		#'what'syourbadassidea.
	- Is there going to be impassable terrain? e.g. water. if so how will this be represented. Could make a class Thing() that has an int x int y bool solid. Unit would be a subclass of this.
		#doesjavascriptimplementinheritance
	- The mouse event registers when it's a mouseup event. This is different to mousedown and click. mousedown is obvious, click is only when both mouseup and mousedown happen in the same location.
		#youonlyhavetoreleasethemouseonaunittoselectit
	- What about undoing a move? To do this you'd need to store the last move made in a variable or something. Or write to some sort of log everytime an action is made, that idea sounds sort of cool. and could enable shit like replays which would be badman. Also unlimited move undoing.
		#'badmanreplaysandwhat.
		
	


Website/Javascript/unit.js
////Classes
function Unit(xc, yc, hp, sprt, drw)
{
	this.x = xc;
	this.y = yc;
	this.health = hp;
	this.sprite = sprt;
	this.draw = drw;
	this.speed = 5; //How many spaces it can move per turn
	this.solid = true;
	this.clickable = true;
	this.current = false;
	
	this.setCurrent = function()
	{
		this.current = true;
		this.sprite = 1;
	}
	this.setNotCurrent = function()
	{
		this.current = false;
		this.sprite = 0;
	}
	this.isCurrent = function()
	{
		return this.current;
	}
	
	this.moveRelative = function(xdelt, ydelt)// called by moveunitrelative by playermove
	{
		this.x += xdelt;
		this.y += ydelt;
	}
	
	this.moveAbsolute = function(x, y) // called by moveunit by playermoveabsolute
	{
		this.x = x;
		this.y = y;
	}
	
	this.playerMove = function(xdelt, ydelt) // relative e.g. (0,1) will go up 1 grid square
	{
		if (validMove(this.x + xdelt, this.y + ydelt)) // check grid to see if legit							
		{	
			moveUnitRelative(this, xdelt, ydelt);
			// OR moveUnit(this,x+xdelt,y+ydelt);
		}
	}
	
	this.playerMoveAbsolute = function(x, y) // Absolute e.g. (0,1) will go up 1 grid square
	{
		if (validMove(x,y)) // check grid to see if legit							
		{	
			moveUnit(this, x, y);
		}
	}
	
	this.validMoveRelative = function(xdelt, ydelt)	// relative - input grid location not pixel location
	{
		var tempx =this.x + xdelt;
		var tempy = this.y + ydelt;
		if (tempx >= 0 && tempy >= 0 &&  tempx < maxX && tempy < maxY)
		{
			if (!occupied(tempx, tempy))
			{
				return true;
			}
		}
		return false;
	}
	
	this.setSolid = function()
	{
		this.solid = true;
	}
	
	this.setUnSolid = function()
	{
		this.solid = false;
	}
}

Website/Javascript/drawer.js
////Drawing functions & vars
var imageList = new Array();
var showOccupied = false;

function drawToBuffer()
{
	drawBackground();
	drawSprites();
	if (showNotOccupied)
	{
		drawNotOccupied();
	}
	if(unitMenuOpen())
	{
		drawUnitMenu();
	}
}

function drawBackground()
{
	buffer.strokeStyle = "rgb(255, 255, 255)";
    buffer.fillStyle = "rgb(255, 255, 255)";
	buffer.fillRect(0,0,_buffer.width,_buffer.height) ;
}

function drawSprites()
{
	for(var i = 0; i < unitList.length; i++)
	{
		var workingUnit = unitList[i];
		if(workingUnit.draw)
		{
			buffer.drawImage(imageList[workingUnit.sprite], workingUnit.x*gridPixel, workingUnit.y*gridPixel);
		}
	}
}

function drawUnitMenu()
{
	buffer.fillStyle = "rgb(0,0,255)";
	buffer.fillRect(unitMenu.x*gridPixel, unitMenu.y*gridPixel, unitMenu.width, unitMenu.height); //Draw the menu
	for(var i = 0; i < unitMenu.options.length; i++)
	{
		var workingOption = unitMenu.options[i];
		buffer.fillStyle = "rgb(255,0,0)";
		buffer.fillRect(workingOption.x, workingOption.y, workingOption.width, workingOption.height);
		buffer.font="14px sans-serif";
		buffer.fillStyle = "rgb(0,0,0)";
		buffer.fillText(workingOption.text, workingOption.x, workingOption.y+workingOption.height-4, workingOption.width);
	}
}

function drawNotOccupied()
{
	for (var i = 0; i < maxX; i++)
	{
		for (var j = 0; j <maxY; j++)
		{
			if (!occupied(i,j))
			{
				buffer.fillStyle = "blue";
				buffer.fillRect(i*gridPixel, j*gridPixel, gridPixel, gridPixel);
			}
		}
	}
}

function check(ctrl)
{
	//get the state of the check box
	if (ctrl.checked == true) 
	{
		showNotOccupied = true;
	}
	else
	{
		showNotOccupied = false;
	}
}

var spriteDirectory = "Resources/Sprites/";

function addImage(fname)
{
	var temp = new Image();
	temp.src = spriteDirectory+fname;
	imageList = imageList.concat([temp]);
}


Website/Javascript/unitmenu.js
var optionHeight = 20; //How tall each menu option is
var menuWidth = 40;

function UnitMenu(x, y, move)
//X & Y are grid position determining where the menu is drawn.  
//All other vars are booleans determining if that part should be drawn
{
	this.x = x;
	this.y = y;
	this.move = move;
	this.options = new Array();
	
	if(this.move)
	{
		this.options = this.options.concat([new menuOption(this, 'Move')]);
	}
	
	this.width = menuWidth;
	this.height = this.options.length*optionHeight;
}

function menuOption(u, text)
//X & Y here are measured in pixel space because they're drawn sub-grid
{
	this.x = u.x*gridPixel+2; //Padding
	this.y = u.y*gridPixel+u.options.length*optionHeight+2; //Padding
	this.width = menuWidth-4;
	this.height = optionHeight-4;
	this.text = text;
	this.rgb = 'rgb(255,0,0)';
	
	this.action = function()
	{
		alert(this.text);
	}
}

Website/Javascript/gamelogic.js
////Game Logic
//Movement checking
var unitGrid;
var unitMenu = null; //Menu to be drawn

gridClicked = function(x, y)
{
	if (yourTurn())
	{
		var clickedUnitIndex
		var found;
		findUnit:
		for(var i = 0; i < unitList.length; i++)
		{
			var workingUnit = unitList[i];
			if(workingUnit.clickable && workingUnit.x == x && workingUnit.y == y)
			{
				clickedUnitIndex = i;
				found = true;
				break findUnit;
			}
		}
		if (found)
		{
			if(clickedUnitIndex == controlledUnit)
			{
				var workingUnit = unitList[controlledUnit];
				unitMenu = new UnitMenu(workingUnit.x+1, workingUnit.y, true);
				changeState(gameState.UNITMENUOPEN);
				clickedUnit(clickedUnitIndex);
			}
			else
			{clickedUnit(clickedUnitIndex);}
		}
	}
	else if (addingUnit())
	{
		if (!occupied(x,y))
		{addUnit(new Unit(x, y, 10, 0, true));}
	}
	else if(unitMenuOpen())
	{
		changeState(gameState.YOURTURN);
	}
}

menuClicked = function(x, y)
//USES PIXEL SPACE
{
	if(x >= unitMenu.x*gridPixel && x <= (unitMenu.x*gridPixel)+unitMenu.width && y >= unitMenu.y*gridPixel && y <= (unitMenu.y*gridPixel)+unitMenu.height)
	{
		var workingOption = findMenuOption(x, y);
		if(workingOption != null)
		{
			workingOption.action();
		}
	}
	
	else
	{
		changeState(gameState.YOURTURN);
	}
}

findMenuOption = function(x, y)
{
	var ret = null;
	var optionArray = unitMenu.options;
	for(var i = 0; i < optionArray.length; i++)
	{
		var workingOption = optionArray[i];
		if(x >= workingOption.x && x <= workingOption.x+workingOption.width && y >= workingOption.y && y <= workingOption.y+workingOption.height)
		{
			ret = workingOption;
			break;
		}
	}
	return ret;
}

setNoCurrent = function()
{
	findCurrent:
	for(var i = 0; i < unitList.length; i++)
	{
		var workingUnit = unitList[i];
		if (workingUnit.isCurrent())
		{
			unitList[i].setNotCurrent();
			break findCurrent;
		}
	}
	controlledUnit = null;
}

clickedUnit = function(i)
{
	if (controlledUnit != null)
	{unitList[controlledUnit].setNotCurrent();}
	controlledUnit = i;
	unitList[controlledUnit].setCurrent();
}
occupied = function(x, y)	// absolute, input grid location not pixel location
{
	if (unitGrid[unitGridArrayConvert(x,y)] == undefined)
	{
		return false;
	}
	else if (unitGrid[unitGridArrayConvert(x,y)].solid == true)
	{
		return true;
	}
	return false;
}

unitGridArrayConvert = function (x, y)
{
	return y*(maxX) + x;
} 
	
validMove = function(x, y)	// absolute - input grid location not pixel location
{
	if (x >= 0 && y >= 0 && x < maxX && y < maxY)
	{
		if (!occupied(x, y))
		{
			return true;
		}
	}
	return false;
}

////Unit manipulation functions
function addUnit(u)
{
	unitList = unitList.concat([u]);
	setGridSquare(u, u.x, u.y);
}

function moveUnit(u, x, y)
{
	emptyGridSquare(u.x, u.y);
	setGridSquare(u, x, y);
	u.moveAbsolute(x,y);
}

function moveUnitRelative(u, xdelt, ydelt)
{
	emptyGridSquare(u.x, u.y);
	setGridSquare(u, u.x+xdelt, u.y+ydelt);
	u.moveRelative(xdelt,ydelt);
}

function setGridSquare(u, x, y)
//This function and emptyGridSquare eventually need to deal with previously occupied squares
{
	unitGrid[unitGridArrayConvert(x, y)] = u;
}

function emptyGridSquare(x, y)
{
	unitGrid[unitGridArrayConvert(x, y)] = null;
}

Website/Javascript/main.js
var _canvas;
var _buffer;
var canvas; //AKA: Context
var buffer; //AKA: Buffer Context
var showNotOccupied = false;

var gridPixel = 32;
var maxX;
var maxY;
var controlledUnit = null; //index
var currentState;

var gameState = {
  YOURTURN: 	{value: 0, name: "Your Turn"}, 
  ADDUNIT:	 	{value: 1, name: "Add Unit"}, 
  UNITMENUOPEN:  {value: 2, name: "Unit Menu Open"},
};

function Game()
{
    this.gameLoop = null;
    var self = this;

    
    this.Init = function()
	{
		_canvas = document.getElementById('canvas');
		if (_canvas && _canvas.getContext)
		{
			canvas = _canvas.getContext('2d');
			
			_buffer = document.createElement('canvas');
			_buffer.width = _canvas.width;
			_buffer.height = _canvas.height;
			buffer = _buffer.getContext('2d');
			
			buffer.strokeStyle = "rgb(255, 255, 255)";
			buffer.fillStyle = "rgb(255, 255, 255)";
			buffer.font = "bold 25px sans-serif";
		}
		maxX = _buffer.width/gridPixel;
		maxY = _buffer.height/gridPixel;
		unitGrid = new Array(maxX, maxY);
		loadImages();
	}
    
    this.Run = function()
	{
		addUnit(new Unit(0,0,10,0,true));
		changeState(gameState.YOURTURN);
		addUnit(new Unit(1,1,10,0,true));
        if(canvas != null)
		{
            self.gameLoop = setInterval(self.Loop, 50);
        }
    }
    
    this.Update = function()
	{
    }
    
    this.Draw = function()
	{
        buffer.clearRect(0, 0, _buffer.width, _buffer.height);
        canvas.clearRect(0, 0, _canvas.width, _canvas.height);
        
        //Draw Code
		drawToBuffer();
        canvas.drawImage(_buffer, 0, 0);
    }
    
    this.Loop = function()
	{
        self.Update();
        self.Draw();    
    }
}

function changeState(newstate)
{
	if (currentState != newstate)
	{
		if (yourTurn())
		{
			setNoCurrent();
		}
		else if (addingUnit())
		{
			showNotOccupied = false;
		}
		currentState = newstate;
		if (yourTurn())
		{
		
		}
		else if (addingUnit())
		{
			showNotOccupied = true;
		}
	}
}
	
//Vars Used Below This Line
var unitList = new Array();

//Loading functions
var imageFnameList = ['square.png','highlightedsquare.png'];

function loadImages()
{
	for(var i = 0; i < imageFnameList.length; i++)
	{
		addImage(imageFnameList[i]);
	}
}

yourTurn = function()
{
	if (currentState == gameState.YOURTURN)
	{return true;}
	return false;
}
addingUnit = function()
{
	if (currentState == gameState.ADDUNIT)
	{return true;}
	return false;
}
unitMenuOpen = function()
{
	if(currentState == gameState.UNITMENUOPEN)
	{return true;}
	return false;
}
////Input
window.addEventListener('keydown', function(event) 
{	// USE TO FIND KEYCODES http://asquare.net/javascript/tests/KeyCode.html
	if (yourTurn() && controlledUnit != null)
	{
		switch (event.keyCode)  // 
		{
			case 37: // Left
				unitList[controlledUnit].playerMove(-1,0);
				break;

			case 38: // Up	
				unitList[controlledUnit].playerMove(0,-1);
				break;

			case 39: // Right
				unitList[controlledUnit].playerMove(1,0);
				break;

			case 40: // Down
				unitList[controlledUnit].playerMove(0,1);
				break;
			case 65: // a
				if (controlledUnit < unitList.length-1) 
				{controlledUnit++;}
				break;
			case 83: // s
				if (controlledUnit > 0)
				{controlledUnit--;}
				break;
			case 32: // space
				changeState(gameState.ADDUNIT);
				break;
		}
	}
	else if (addingUnit())
	{
		switch (event.keyCode)  // 
		{
			case 32: // space
			changeState(gameState.YOURTURN);
			break;
		}
	}
	else
	{
		switch (event.keyCode)  // 
		{
			case 32: // space
			changeState(gameState.ADDUNIT);
			break;
		}
	}
}, false);

function findPos(obj)  // STOLEN http://www.quirksmode.org/js/findpos.html
{
	var curleft = curtop = 0;
	if (obj.offsetParent) 
	{
		do 
		{
			curleft += obj.offsetLeft;
			curtop += obj.offsetTop;
		}	 while (obj = obj.offsetParent);
	}
	return [curleft,curtop];
}

window.addEventListener('mouseup', function(event)
{
	if (event.pageX == null)	// STOLEN http://unixpapa.com/js/mouse.html
   {
      // IE case
      var d= (document.documentElement && 
              document.documentElement.scrollLeft != null) ?
             document.documentElement : document.body;
      docX= event.clientX + d.scrollLeft;
      docY= event.clientY + d.scrollTop;
   }
   else
   {
      // all other browsers
      docX= event.pageX;
      docY= event.pageY;
   }
										// END STOLEN
   var elemPos = findPos(document.getElementById('canvas'));
   mousex = docX - elemPos[0];		//Where the mouse was clicked relative to canvas(0,0)
   mousey = docY - elemPos[1];		//Same
   handleMouseClick(mousex,mousey);
 }, false);
 
handleMouseClick = function(x, y)
{
	if(unitMenuOpen())
	{menuClicked(x, y);}
	else
	{gridClicked(Math.floor(x/gridPixel),Math.floor(y/gridPixel));}
}
